proposals/0001-dependency-injection.md

```diff
@@ -0,0 +1,527 @@
+
+### Contents
+
+1. [Summary](#summary)
+2. [Motivation](#motivation)
+3. [Current Solutions](#current-solutions)
+   1. [Koin](#koin)
+   2. [Kodein](#kodein)
+   3. [Spring](#spring)
+   4. [Dagger, Micronaut, etc.](#dagger-micronaut-etc)
+4. [Design Overview](#design-overview)
+5. [Design Details](#design-details)
+   1. [Declaration](#declaration)
+   2. [Naming](#naming)
+   3. [Resolution](#resolution)
+      1. [Naming](#naming-1)
+6. [Technical Details](#technical-details)
+   1. [Instantiation by Reflection](#instantiation-by-reflection)
+   2. [Validation](#validation)
+      1. [Compile-time Validation](#compile-time-validation)
+   3. [Dependency lifecycles](#dependency-lifecycles)
+   4. [Performance](#performance)
+   5. [Resolving from Configuration](#resolving-from-configuration)
+   6. [Extensibility](#extensibility)
+   7. [Testing](#testing)
+7. [Drawbacks](#drawbacks)
+8. [Advantages](#advantages)
+9. [Open Questions](#open-questions)
+10. [Future Directions](#future-directions)
+
+<hr />
+
+# Summary
+[summary]: #summary
+
+We describe a general solution for handling [dependency injection (DI)](https://en.wikipedia.org/wiki/Dependency_injection)
+in Ktor server applications.  Our intent is to provide a simple, performant interface, while avoiding the drawbacks that
+that can arise when working with the automatic configuration of dependencies.
+
+# Motivation
+[motivation]: #motivation
+
+We receive many support requests from users having difficulty getting started with dependency injection in Ktor,
+especially when coming from a background of using frameworks like Spring Boot.  There are several third party libraries
+that can be used with Ktor to provide dependency injection; however, the process of selection, integration, and maintenance,
+creates a barrier to adoption.  Furthermore, our documentation for more advanced use-cases can suffer from a lack of
+DI support, which exacerbates challenges with learning the framework.
+
+Our goal for this tooling is to provide a clear idiomatic approach to declaring and resolving dependencies in Ktor applications.
+The resulting API should work "out of the box" without any external dependencies, but should also provide
+some integration paths for existing projects to leverage their current DI solutions for declaring dependencies.
+
+# Current Solutions
+[current-solutions]: #current-solutions
+
+### Koin
+[koin]: #koin
+
+[Koin](https://insert-koin.io/) is a lightweight implementation, which avoids using reflection and favors idiomatic Kotlin.
+
+There is also a Ktor plugin provided in the project generator currently.
+
+The API looks like this:
+
+```kotlin
+// installation
+fun Application.installDependencies() {
+    koin {
+        modules(module {
+            single { Service() }
+            factory(named("key")) { Repository(get()) }
```

Comment:
In Koin you can also [pass a constructor](https://github.com/InsertKoinIO/koin/blob/fa85a195e748ef14c5a3c5abe474f2bffebeb579/projects/core/koin-core/src/commonMain/kotlin/org/koin/core/module/dsl/FactoryOf.kt#L41-L44) instead of lambda. All parameters will be retrieved using `get()`:
```
factoryOf(::Repository)
```
This solution is a compromise between manual and reflective instantiation.

Reply:
Oh, I like that!

Reply:
If I can add some clarification here: no reflection, it's a (dirty) trick to match the signature and fill in the needed dependencies.

Reply:
> If I can add some clarification here: no reflection, it's a (dirty) trick to match the signature and fill in the needed dependencies.

Thank you for the clarification! Yes, I called it "a compromise" because it is as fast as manual instantiation but lacks some features of reflective instantiation. Though, I can name only one missing feature - named instances aren’t supported.

Reply:
yes, this is a point where I would need a compiler plugin or something here

Reply:
Yeah, I don't think it's possible to get around the use of annotations + reflection for named instances when injecting constructors.

Reply:
In spring framework, previous to annotations, we had to specify it manually